src/osd/PrimaryLogPG.cc

```diff
@@ -15109,23 +15106,15 @@ int PrimaryLogPG::rep_repair_primary_object(const hobject_t& soid, OpContext *ct
   waiting_for_unreadable_object[soid].push_back(op);
   op->mark_delayed("waiting for missing object");
 
-  if (!eio_errors_to_process) {
-    eio_errors_to_process = true;
-    ceph_assert(is_clean());
-    state_set(PG_STATE_REPAIR);
-    state_clear(PG_STATE_CLEAN);
-    queue_peering_event(
-        PGPeeringEventRef(
-	  std::make_shared<PGPeeringEvent>(
-	  get_osdmap_epoch(),
-	  get_osdmap_epoch(),
-	  PeeringState::DoRecovery())));
-  } else {
-    // A prior error must have already cleared clean state and queued recovery
-    // or a map change has triggered re-peering.
-    // Not inlining the recovery by calling maybe_kick_recovery(soid);
-    dout(5) << __func__<< ": Read error on " << soid << ", but already seen errors" << dendl;
-  }
+  ceph_assert(is_clean());
```

Comment:
The rep_repair_primary_object() can be called by multiple clients read requests getting EIO.

This assert will trip when rep_repair_primary_object() is called a second time because state_clear(PG_STATE_CLEAN) will have been called by the first request.

Reply:
> This assert will trip when rep_repair_primary_object() is called a second time because state_clear(PG_STATE_CLEAN) will have been called by the first request.

@dzafman Did you see https://github.com/ceph/ceph/blob/master/src/osd/PrimaryLogPG.cc#L15088

All following ops that trigger a repair will go to the waiting_for_clean_to_primary_repair queue.

Reply:
@xiexingguo OK, the clean bit is acting like the eio_errors_to_process flag.  It is unfortunate that if recovery is happening we can't  handle multiple errors instead of having to go in and out of recovery.  In the future, what if we want to not block repair on !is_clean()?  Maybe if backfilling or Recovering is set we can go forward even if !is_clean().  How would Crimson handle the case that we need that flag again later?

Reply:
Does it matter that eio_errors_to_process is a PG member not really a global?

Reply:
@dzafman Why would we want to go forward if the PG is already in backfilling/recovering state? In `rep_repair_primary_object()` after the point in which you check `is_clean()`, you immediately schedule a recovery of an object, nothing else is happening in between (other than changing the state).

Reply:
@AmnonHanuhov Because if something is added to the pg log and we are already in backfill or recovery, it could be handled before we could get out of backfill/recovery.  Otherwise, each individual read error has to enter recovery and leave recovery.  In the scenario where a portion of a disk gets damaged multiple read errors could occur though I'm not a hardware expert. We would change the test as follows:

```
 if (!is_clean() && !is_backfilling() && !is_recovering()) {
           block_for_clean(soid, op);  // Fix function name?
          return -EAGAIN;
  }
```

And in `rep_repair_primary_object()` the assert would look like:

```
ceph_assert(is_clean() || is_backfilling() || is_recovery());
```

These change would allow us to handle all read errors that occur while clear OR during backfill/recovery and when everything is done the PG goes clean.  If this works and I changed it in master what would Crimson do?  I'd want to add back the `eio_errors_to_process`.

Reply:
Now that I've thought about and worked out the change the clean bit can still be used to get rid of  `eio_errors_to_process`!

The change below hasn't been tested.

```
--- a/src/osd/PrimaryLogPG.cc
+++ b/src/osd/PrimaryLogPG.cc
@@ -15092,8 +15092,8 @@ int PrimaryLogPG::rep_repair_primary_object(const hobject_t& soid, OpContext *ct
   dout(10) << __func__ << " " << soid
           << " peers osd.{" << get_acting_recovery_backfill() << "}" << dendl;

-  if (!is_clean()) {
-    block_for_clean(soid, op);
+  if (!is_clean() && !is_backfilling && !is_recovering()) {
+    block_for_clean(soid, op);  // XXX: Fix function name?
     return -EAGAIN;
   }

@@ -15115,9 +15115,7 @@ int PrimaryLogPG::rep_repair_primary_object(const hobject_t& soid, OpContext *ct
   waiting_for_unreadable_object[soid].push_back(op);
   op->mark_delayed("waiting for missing object");

-  if (!eio_errors_to_process) {
-    eio_errors_to_process = true;
-    ceph_assert(is_clean());
+  if (is_clean()) {
     state_set(PG_STATE_REPAIR);
     state_clear(PG_STATE_CLEAN);
     queue_peering_event(
@@ -15127,6 +15125,8 @@ int PrimaryLogPG::rep_repair_primary_object(const hobject_t& soid, OpContext *ct
          get_osdmap_epoch(),
          PeeringState::DoRecovery())));
   } else {
+    // Set repair in caes we are the first read error and we happen to be backfilling or recovering
+    state_set(PG_STATE_REPAIR);
     // A prior error must have already cleared clean state and queued recovery
     // or a map change has triggered re-peering.
     // Not inlining the recovery by calling maybe_kick_recovery(soid);--- ```